### PR TITLE
`kubelet`: Allow changing pod log directory

### PR DESCRIPTION
**Pull request by danilo-gemoli**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
https://github.com/kubernetes/kubernetes/issues/98473

#### Does this PR introduce a user-facing change?

```release-note
Add a new field `PodLogsRootDirectory` to the `KubeletConfiguration` that allow the user to set the pod logs root directory.  
It defaults to `/var/log/pods` if unset, as it was before.
```

FYI: @lorenz @pacoxu

/label tide/merge-method-squash

### CODE REVIEW

**Comment by danilo-gemoli** [comment]
There are 14 commits but it's not that scary as it was just me taking little steps. I can eventually squash them all into a single one, even though [tide](https://github.com/kubernetes/test-infra/tree/master/prow/cmd/tide) should do that, hopefully.

**Comment by lorenz** [comment]
Thanks, I'm going to try to test this instead of our patch if I can easily rebase it on top of 1.28.

**Comment by kannon92** [comment]
> There are 14 commits but it's not that scary as it was just me taking little steps. I can eventually squash them all into a single one, even though [tide](https://github.com/kubernetes/test-infra/tree/master/prow/cmd/tide) should do that, hopefully.

I would suggest just squashing them.

**Comment by kannon92** [comment]
@danilo-gemoli Thanks for this change. This sounds like a useful change. I would ask that we make sure to update the kubernetes docs to reflect this change.

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go [hunk 1]
```diff
@@ -253,6 +254,7 @@ func TestSetDefaultsKubeletConfiguration(t *testing.T) {
 				MemoryThrottlingFactor:          utilpointer.Float64(0),
 				RegisterNode:                    utilpointer.Bool(false),
 				LocalStorageCapacityIsolation:   utilpointer.Bool(false),
+				PodLogsRootDirectory:            "",
```
Comment (kannon92) [inline pkg/kubelet/apis/config/v1beta1/defaults_test.go:257]: I'd like to see one more test case where you set the `PodLogsRootDirectory` to non default and empty. I know the code will probably work but we should cover that case also.

File: pkg/kubelet/apis/config/types.go [hunk 2]
```diff
@@ -484,6 +484,11 @@ type KubeletConfiguration struct {
 	// If not specified the default value is ContainerRuntimeEndpoint
 	// +optional
 	ImageServiceEndpoint string
+
+	// PodLogsRootDirectoryPodLogsRootDirectory is the root directory for pod logs.
```
Comment (kannon92) [inline pkg/kubelet/apis/config/types.go:488]: What does "root directory" mean here?
```suggestion
	// PodLogsRootDirectoryPodLogsRootDirectory is the directory for pod logs.
```
And please make sure to update the staging api comment also if you add this change.

**Review by kannon92 — COMMENTED** [review]
https://github.com/kubernetes/kubernetes/pull/112957 seems to resolve this issue. If that author is responsive, I don't see us going forward with this PR.

File: pkg/kubelet/apis/config/v1beta1/defaults_test.go (hunk 1 above)
Comment (danilo-gemoli) [reply pkg/kubelet/apis/config/v1beta1/defaults_test.go:257]: Yes it's reasonable. Done.

**Comment by kannon92** [comment]
> #112957 seems to resolve this issue. If that author is responsive, I don't see us going forward with this PR.

@mxpv responded to this PR so I don't see any reason to continue with this one. There are two items I think we should add for #112957, but could be done as followups. It would be nice to see some e2e tests for pod log rotation with a separate location and we should update our documentation to mention that you can change the pod log directory. Maybe you could help with that?

**Comment by danilo-gemoli** [comment]
Of course I can help. As @kannon92 mentioned, this PR is being superseded by #112957, we can then
/close

**Pull request closed without merging**